crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs

```diff
@@ -0,0 +1,254 @@
+            }
+            Stmt::AugAssign(ast::StmtAugAssign { target, value, .. }) => {
+                self.is_subcript_modified =
+                    check_target_for_assignment(target, self.sequence_name, self.index_name);
+                self.visit_expr(value);
+            }
+            _ => visitor::walk_stmt(self, stmt),
+        }
+    }
+}
+
+/// PLR1736
+pub(crate) fn unnecessary_list_index_lookup(checker: &mut Checker, stmt_for: &StmtFor) {
+    let Some((sequence, index_name, value_name)) =
+        enumerate_items(checker, &stmt_for.iter, &stmt_for.target)
+    else {
+        return;
+    };
+
+    let mut visitor = SubscriptVisitor::new(&sequence, &index_name);
+
+    visitor.visit_body(&stmt_for.body);
+    visitor.visit_body(&stmt_for.orelse);
+
+    for range in visitor.diagnostic_ranges {
+        let mut diagnostic = Diagnostic::new(UnnecessaryListIndexLookup, range);
+
+        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+            value_name.clone(),
+            range,
+        )));
+
+        checker.diagnostics.push(diagnostic);
+    }
+}
+
+/// PLR1736
+pub(crate) fn unnecessary_list_index_lookup_comprehension(checker: &mut Checker, expr: &Expr) {
+    match expr {
+        Expr::GeneratorExp(ast::ExprGeneratorExp {
+            elt, generators, ..
+        })
+        | Expr::DictComp(ast::ExprDictComp {
+            value: elt,
+            generators,
+            ..
+        })
+        | Expr::SetComp(ast::ExprSetComp {
+            elt, generators, ..
+        })
+        | Expr::ListComp(ast::ExprListComp {
+            elt, generators, ..
+        }) => {
+            for comp in generators {
+                if let Some((sequence, index_name, value_name)) =
+                    enumerate_items(checker, &comp.iter, &comp.target)
+                {
+                    let mut visitor = SubscriptVisitor::new(&sequence, &index_name);
+
+                    visitor.visit_expr(elt.as_ref());
+
+                    for range in visitor.diagnostic_ranges {
+                        let mut diagnostic = Diagnostic::new(UnnecessaryListIndexLookup, range);
+
+                        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                            value_name.clone(),
+                            range,
+                        )));
+
+                        checker.diagnostics.push(diagnostic);
+                    }
+                }
+            }
+        }
+        _ => (),
+    }
+}
+
+fn enumerate_items(
+    checker: &mut Checker,
+    call_expr: &Expr,
+    tuple_expr: &Expr,
+) -> Option<(String, String, String)> {
+    let Expr::Call(ast::ExprCall {
+        func, arguments, ..
+    }) = call_expr
+    else {
+        return None;
+    };
+
+    // Check that the function is the `enumerate` builtin.
+    let Expr::Name(ast::ExprName { id, .. }) = func.as_ref() else {
+        return None;
+    };
+    if id != "enumerate" {
+        return None;
+    };
```

Comment:
Should this use `checker.semantic().resolve_call_path(func)`?

Reply:
Good change, I've also added detection for `builtins.enumerate` along with it, thanks!